worlds/ahit/DeathWishRules.py

```diff
@@ -0,0 +1,533 @@
+    full_clear = world.multiworld.get_location(f"{name} - All Clear", world.player)
+
+    if name == "Zero Jumps":
+        add_rule(main_objective, lambda state: get_zero_jump_clear_count(state, world) >= 1)
+        add_rule(full_clear, lambda state: get_zero_jump_clear_count(state, world) >= 4
+                 and state.has("Train Rush (Zero Jumps)", world.player) and can_use_hat(state, world, HatType.ICE))
+
+        # No Ice Hat/painting required in Expert for Toilet Zero Jumps
+        # This painting wall can only be skipped via cherry hover.
+        if get_difficulty(world) < Difficulty.EXPERT or world.options.NoPaintingSkips.value == 1:
+            set_rule(world.multiworld.get_location("Toilet of Doom (Zero Jumps)", world.player),
+                     lambda state: can_use_hookshot(state, world) and can_hit(state, world)
+                     and has_paintings(state, world, 1, False))
+        else:
+            set_rule(world.multiworld.get_location("Toilet of Doom (Zero Jumps)", world.player),
+                     lambda state: can_use_hookshot(state, world) and can_hit(state, world))
+
+        set_rule(world.multiworld.get_location("Contractual Obligations (Zero Jumps)", world.player),
+                 lambda state: has_paintings(state, world, 1, False))
+
+    elif name == "Snatcher's Hit List":
+        add_rule(main_objective, lambda state: state.has("Mafia Goon", world.player))
+        add_rule(full_clear, lambda state: get_reachable_enemy_count(state, world) >= 12)
+
+    elif name == "Camera Tourist":
+        add_rule(main_objective, lambda state: get_reachable_enemy_count(state, world) >= 8)
+        add_rule(full_clear, lambda state: can_reach_all_bosses(state, world)
+                 and state.has("Triple Enemy Photo", world.player))
+
+    elif "Snatcher Coins" in name:
+        coins: List[str] = []
+        for coin in required_snatcher_coins[name]:
+            coins.append(coin)
+            add_rule(full_clear, lambda state: state.has(coin, world.player))
+
+        # any coin works for the main objective
+        add_rule(main_objective, lambda state: state.has(coins[0], world.player)
+                 or state.has(coins[1], world.player)
+                 or state.has(coins[2], world.player))
+
+
+def get_zero_jump_clear_count(state: CollectionState, world: "HatInTimeWorld") -> int:
+    total = 0
+
+    for name in act_chapters.keys():
+        n = f"{name} (Zero Jumps)"
+        if n not in zero_jumps:
+            continue
+
+        if get_difficulty(world) < Difficulty.HARD and n in zero_jumps_hard:
+            continue
+
+        if get_difficulty(world) < Difficulty.EXPERT and n in zero_jumps_expert:
+            continue
+
+        if not state.has(n, world.player):
+            continue
+
+        total += 1
+
+    return total
+
+
+def get_reachable_enemy_count(state: CollectionState, world: "HatInTimeWorld") -> int:
```

Comment:
Unfortunately every enemy type has to be a separate item because the Camera Tourist and Snatcher's Hit List require you to snap a picture of or kill each type of enemy. I'm not entirely sure what you mean by overriding world.collect and remove though.

Reply:
https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/messenger/__init__.py#L362

Reply:
So this is basically merging all of the items into a singular one?

Reply:
basically, yeah. so every time you collect any enemy you collect a fake "Enemy" item to the state, and just check if you have enough of that item instead of checking for all of the other items

Reply:
def collect(self, state: "CollectionState", item: "Item") -> bool:
        old_count: int = state.count(item.name, self.player)
        change = super().collect(state, item)
        if change and old_count == 0:
            if item.name in zero_jumps.keys():
                state.prog_items[self.player]["Zero Jumps"] += 1
            elif item.name in hit_list.keys():
                if item.name not in bosses:
                    state.prog_items[self.player]["Enemy"] += 1
                else:
                    state.prog_items[self.player]["Boss"] += 1

        return change

    def remove(self, state: "CollectionState", item: "Item") -> bool:
        old_count: int = state.count(item.name, self.player)
        change = super().collect(state, item)
        if change and old_count == 1:
            if item.name in zero_jumps.keys():
                state.prog_items[self.player]["Zero Jumps"] -= 1
            elif item.name in hit_list.keys():
                if item.name not in bosses:
                    state.prog_items[self.player]["Enemy"] -= 1
                else:
                    state.prog_items[self.player]["Boss"] -= 1

        return change
        
 So something like this?

Reply:
yeah that looks right